### ISSUE TITLE
用自己的数据集训练为什么效果变得很差

### ISSUE THREAD

**Issue by Wqf112** [opendatalab/DocLayout-YOLO#164]
您好！

我用的是doclayout_yolo_docstructbench_imgsz1280_2501.pt模型在x-anylabeling上打标训练了1500张图片，然后用训练自己的模型发现效果很差，请问是什么原因

这是训练出来的指标
<img width="957" height="247" alt="Image" src="https://github.com/user-attachments/assets/658023a8-e655-431e-9ab2-4e7b56bffbe3" />

这是train的代码
`model_path = '/home/wqf/model/doclayout_yolo_docstructbench_imgsz1280_2501.pt'  # 权重路径`
`model = YOLOv10(model_path)`
`results = model.train(data='/home/wqf/data/特殊研报_3_1_第一批_train_1/segmentation.yaml',epochs=200,name='doclayout_yolo_1280_3',# 初始设置较大的训练轮数
    imgsz=1280,
    batch=4,
    patience=30,
    lr0=0.0002,
    optimizer='Adam',
    freeze=20,
    device='1'
    )`

**Comment by Wqf112** [comment]
这是我训练的模型在x-anylabeling上运行的结果，模型把plain_text划分为了abandon，把figure划分为了title，而用doclayout_yolo_docstructbench_imgsz1280_2501可以精准划分，我的数据集没有问题，请问是训练轮次少了还是参数设置有问题

<img width="687" height="805" alt="Image" src="https://github.com/user-attachments/assets/c57949ac-54a2-42b3-89a8-b8247c3cfdb7" />

**Comment by Wqf112** [comment]
@JulioZhao97

**Comment by JulioZhao97** [comment]
@Wqf112 感觉是参数设置有问题，上面那个训练出来的指标那里，mAP是没问题的，all的mAP能到93.3；然后您的可视化的图片那里，感觉是哪里参数有问题类别混乱了，因为所有类别都是错的，但是框没问题，您可以再检查一下训练数据和推理的过程

**Comment by Wqf112** [comment]
啊，怎么检查呢，这是yaml配置

<img width="865" height="597" alt="Image" src="https://github.com/user-attachments/assets/225f2f03-0228-4e94-9032-8ae6e0774069" />
@JulioZhao97

**Comment by Wqf112** [comment]
感觉好像是模型的问题，我用训练出来的模型跑我们公司的代码，标题提取的好乱

这是源pdf，目标是提取出标题

<img width="817" height="781" alt="Image" src="https://github.com/user-attachments/assets/89e19782-2cf6-4a43-aa68-97634e5e9d4e" />

这是提取的效果

<img width="695" height="118" alt="Image" src="https://github.com/user-attachments/assets/1e24d15e-385e-48bb-a0cd-dd30a669f18e" />

1，2标题都没有提取出来

**Comment by Wqf112** [comment]
但是我用doclayout_yolo_docstructbench_imgsz1280_2501模型可以正确提取，用自己的模型就提取的特别乱

**Comment by Wqf112** [comment]
@JulioZhao97

**Comment by JulioZhao97** [comment]
> 啊，怎么检查呢，这是yaml配置
> 
> <img alt="Image" width="865" height="597" src="https://private-user-images.githubusercontent.com/151728310/471346389-225f2f03-0228-4e94-9032-8ae6e0774069.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM2ODQ1MjMsIm5iZiI6MTc1MzY4NDIyMywicGF0aCI6Ii8xNTE3MjgzMTAvNDcxMzQ2Mzg5LTIyNWYyZjAzLTAyMjgtNGU5NC05MDMyLThhZTZlMDc3NDA2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyOFQwNjMwMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNWE3YzMzY2YxYzA2MWM1ZTlmYWI2YmM2MjE0MjY5ZmRhMDhkYmRmNzY4MTlhMWFjOTg4NzcxNDgxMmNkZjMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RGmsBLw8_cfj9jU8-ohXUZT1FzU5QD1OoXyW5IzzNPg"> [@JulioZhao97](https://github.com/JulioZhao97)

这里的model_path能换一个非onnx的模型看一下吗，onnx的模型可能会有点问题

**Comment by Wqf112** [comment]
怎么换呢

**Comment by Wqf112** [comment]
我用doclayout_yolo_docstructbench_imgsz1280_2501.onnx没有问题

**Comment by Wqf112** [comment]
是不是转成yolo数据集的时候出问题了，这是代码

import os
import json
import cv2
import numpy as np
from tqdm import tqdm
from concurrent.futures import ThreadPoolExecutor, as_completed

# 将lableme格式的数据转成可用YOLO训练的数据集的格式

# 定义类别名称
# class_names = ['Figure', 'Question','Table', 'Text', 'Title']
class_names = ['figure', 'table', 'plain_text', 'title', 'abandon', 'figure_caption', 'table_caption' , 'table_footnote', 'formula_caption', 'isolate_formula']
# names: 

# 数据集路径
labelme_json_dir = r'/home/wqf/data/特殊研报_3_1_第一批'
output_dir = r'/home/wqf/data/特殊研报_3_1_第一批_train_1'

# 创建输出文件夹
os.makedirs(output_dir, exist_ok=True)
os.makedirs(os.path.join(output_dir, 'images/train'), exist_ok=True)
os.makedirs(os.path.join(output_dir, 'images/val'), exist_ok=True)
os.makedirs(os.path.join(output_dir, 'labels/train'), exist_ok=True)
os.makedirs(os.path.join(output_dir, 'labels/val'), exist_ok=True)

# 获取所有JSON文件
json_files = [f for f in os.listdir(labelme_json_dir) if f.endswith('.json')]

# 将80%的数据用作训练集，20%用作验证集
train_split = int(0.8 * len(json_files))
train_files = json_files[:train_split]
val_files = json_files[train_split:]

def convert_labelme_to_yolo(json_file, output_image_dir, output_label_dir):
    with open(json_file, 'r') as f:
        data = json.load(f)

    # 去掉特定部分
    new_path = data['imagePath'].replace("..\\", "")
    new_path=new_path.replace(".png",".jpg")
    image_path = os.path.join(os.path.dirname(json_file), new_path)
    image_path = image_path.replace(".png", ".jpg")
    image = cv2.imdecode(np.fromfile(image_path, dtype=np.uint8), cv2.IMREAD_COLOR)
    height, width, _ = image.shape

    # 处理标注信息
    label_file = os.path.join(output_label_dir, os.path.splitext(os.path.basename(json_file))[0] + '.txt')
    with open(label_file, 'w') as f:
        for shape in data['shapes']:
            points = np.array(shape['points'])
            xmin = np.min(points[:, 0])
            xmax = np.max(points[:, 0])
            ymin = np.min(points[:, 1])
            ymax = np.max(points[:, 1])
            try:
                class_id = class_names.index(shape['label'])
            except:
                print(f"label_file: {label_file} 出错！")
                return None

            # YOLO格式：class_id center_x center_y width height
            center_x = (xmin + xmax) / 2 / width
            center_y = (ymin + ymax) / 2 / height
            w = (xmax - xmin) / width
            h = (ymax - ymin) / height
            f.write(f"{class_id} {center_x} {center_y} {w} {h}\n")

    # 保存图像到输出目录
    output_image_path = os.path.join(output_image_dir, os.path.basename(image_path))
    cv2.imwrite(output_image_path, image)
    # cv2.imencode(".png", image)[1].tofile(output_image_path)

def process_files(files, output_image_dir, output_label_dir, progress_bar):
    with ThreadPoolExecutor(max_workers=16) as executor:
        futures = []
        for json_file in files:
            future = executor.submit(convert_labelme_to_yolo, os.path.join(labelme_json_dir, json_file), output_image_dir, output_label_dir)
            futures.append(future)

        for future in as_completed(futures):
            future.result()
            progress_bar.update(1)

# 使用多线程转换训练集和验证集
with ThreadPoolExecutor(max_workers=2) as executor:
    train_progress_bar = tqdm(total=len(train_files), desc='Converting train set')
    val_progress_bar = tqdm(total=len(val_files), desc='Converting val set')

    train_future = executor.submit(process_files, train_files, os.path.join(output_dir, 'images/train'), os.path.join(output_dir, 'labels/train'), train_progress_bar)
    val_future = executor.submit(process_files, val_files, os.path.join(output_dir, 'images/val'), os.path.join(output_dir, 'labels/val'), val_progress_bar)

    # 等待所有任务完成
    for future in as_completed([train_future, val_future]):
        future.result()

    train_progress_bar.close()
    val_progress_bar.close()
@JulioZhao97

**Comment by Wqf112** [comment]
我好像找到问题了，在segmentation.yaml里面设置nc=8会不会有问题呢，打标的标签数量是10，但是实际用到的标签只有8个，所以我在segmentation.yaml里面设置了nc=8，这样会不会有问题
@JulioZhao97

**Comment by Wqf112** [comment]
还是有问题
@JulioZhao97

**Comment by JulioZhao97** [comment]
@Wqf112 这个best_2.onnx是什么模型呢？是你自己训的吗？

**Comment by JulioZhao97** [comment]
@Wqf112 这个我暂时没空帮你弄，明天再看看

**Comment by Wqf112** [comment]
> [@Wqf112](https://github.com/Wqf112) 这个best_2.onnx是什么模型呢？是你自己训的吗？

对，是自己训练的
@JulioZhao97

**Comment by JulioZhao97** [comment]
@JulioZhao97 您用非onnx格式的模型跑一下试试看呢？

**Comment by Wqf112** [comment]
怎么用非onnx格式

**Comment by JulioZhao97** [comment]
就是您不是自己训的模型吗，默认会保存.pth格式的模型（torch格式的），这个找不到吗

**Comment by JulioZhao97** [comment]
@Wqf112 感觉是导出的onnx模型的问题

**Comment by Wqf112** [comment]
我在自已的代码上用的是pt格式的，效果也是一样
@JulioZhao97

**Comment by Wqf112** [comment]
您好，解决了，原来是我最开始的数据集的yaml的标签全反了

**Comment by Wqf112** [comment]
感谢您的耐心解答，祝您生活愉快

**Comment by JulioZhao97** [comment]
> 您好，解决了，原来是我最开始的数据集的yaml的标签全反了

好的
